lib/usd/translators/shading/mtlxMaterialXSurfaceShaderWriter.cpp

```diff
@@ -0,0 +1,516 @@
+            }
+        }
+        if (isNodeGraph) {
+            _AddNodeGraphInput(input, usdNodeGraph, parentPath, targetUfePath, stage);
+        }
+    }
+}
+
+class MtlxMaterialXSurfaceShaderWriter : public UsdMayaShaderWriter
+{
+public:
+    MtlxMaterialXSurfaceShaderWriter(
+        const MFnDependencyNode& depNodeFn,
+        const SdfPath&           usdPath,
+        UsdMayaWriteJobContext&  jobCtx);
+
+    static ContextSupport CanExport(const UsdMayaJobExportArgs&);
+    void                  Write(const UsdTimeCode& usdTime) override;
+};
+
+PXRUSDMAYA_REGISTER_SHADER_WRITER(MaterialXSurfaceShader, MtlxMaterialXSurfaceShaderWriter);
+
+UsdMayaShaderWriter::ContextSupport
+MtlxMaterialXSurfaceShaderWriter::CanExport(const UsdMayaJobExportArgs& exportArgs)
+{
+    if (!exportArgs.exportMaterials) {
+        return ContextSupport::Unsupported;
+    }
+
+    return exportArgs.convertMaterialsTo == TrMtlxTokens->conversionName
+        ? ContextSupport::Supported
+        : ContextSupport::Unsupported;
+}
+
+MtlxMaterialXSurfaceShaderWriter::MtlxMaterialXSurfaceShaderWriter(
+    const MFnDependencyNode& depNodeFn,
+    const SdfPath&           usdPath,
+    UsdMayaWriteJobContext&  jobCtx)
+    : UsdMayaShaderWriter(depNodeFn, usdPath, jobCtx)
+{
+    // The shader writer is being called twice, once for the surface and once for the
+    // displacement, but there is only one material
+    // Skip the second call
+    if (GetUsdStage()->GetPrimAtPath(GetUsdPath())) {
+        return;
+    }
+
+    UsdShadeShader shaderSchema = UsdShadeShader::Define(GetUsdStage(), GetUsdPath());
+    if (!TF_VERIFY(
+            shaderSchema,
+            "Could not define UsdShadeShader at path '%s'\n",
+            GetUsdPath().GetText())) {
+        return;
+    }
+
+    _usdPrim = shaderSchema.GetPrim();
+    if (!TF_VERIFY(
+            _usdPrim,
+            "Could not get UsdPrim for UsdShadeShader at path '%s'\n",
+            shaderSchema.GetPath().GetText())) {
+        return;
+    }
+    auto parentPath = GetUsdPrim().GetParent().GetPath();
+
+    MStatus status;
+    MPlug   childPlug = depNodeFn.findPlug("ufePath", true, &status);
+    MString ufePathString;
+    childPlug.getValue(ufePathString);
+
+    // This is the material node
+    auto ufePath = Ufe::PathString::path(ufePathString.asChar());
+    // This is the document node
+    auto                ufeParentPath = ufePath.pop();
+    Ufe::SceneItem::Ptr sceneItem = Ufe::Hierarchy::createItem(ufeParentPath);
+
+    // Render Document is the MaterialX document
+    childPlug = depNodeFn.findPlug("renderDocument", true, &status);
+    MString renderDocumentString;
+    childPlug.getValue(renderDocumentString);
+
+    auto mtlxDoc = MaterialX::createDocument();
+    readFromXmlString(mtlxDoc, renderDocumentString.asChar());
+
+    // surfaceMaterialNode
+    auto MaterialNode = mtlxDoc->getNode(depNodeFn.name().asChar());
```

Comment:
I try to keep that name in sync with the MaterialX surfacematerial node name, but it can diverge since it must be unique in Maya. The real material node name can be found at the end of the path: `ufePath.back()`.

For example, if you import the same `standard_surface_gold.mtlx` MaterialX document twice, and assign `Gold` from each document on separate spheres, then the first depNodeFn will have `Gold` as name, but the second one will be stuck with `Gold1` since these need to be unique. In both cases the last component of the Ufe path will be correct:
```
|materialXStack1|materialXStackShape1,%standard_surface_gold%Gold
|materialXStack1|materialXStackShape1,%standard_surface_gold2%Gold
```